framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceScenes.java

```diff
@@ -332,17 +381,30 @@ private String getEnclosingClassName(LocalVariableNode localVariableNode) {
      * @param executableElement the ExecutableElement
      * @return the "flatname" of the class enclosing {@code executableElement}
      */
-    private String getEnclosingClassName(ExecutableElement executableElement) {
+    @SuppressWarnings("signature") // https://tinyurl.com/cfissue/3094
+    private @BinaryName String getEnclosingClassName(ExecutableElement executableElement) {
         return ((MethodSymbol) executableElement).enclClass().flatName().toString();
     }
 
     /**
-     * Returns the "flatname" of the class enclosing {@code variableElement}
+     * Updates the metadata stored in AClassWrapper for the given class.
      *
-     * @param variableElement the VariableElement
-     * @return the "flatname" of the class enclosing {@code variableElement}
+     * @param classSymbol the class for which to update metadata
+     * @param aClassWrapper the class representation in which the metadata is to be updated
      */
-    private String getEnclosingClassName(VariableElement variableElement) {
-        return ((VarSymbol) variableElement).enclClass().flatName().toString();
+    private void updateClassMetadata(ClassSymbol classSymbol, AClassWrapper aClassWrapper) {
+        if (classSymbol.isEnum()) {
+            if (!aClassWrapper.isEnum()) {
```

Comment:
If it already is an enum, should this code assert that the given argument is the same as what is supplied?  Or should this code throw an exception because its contract is that it is called just once?
(Same question about the assignment on line 408.)

Reply:
`setTypeElement` (line 408) throws an exception if what's provided isn't what it already has, as does `markAsEnum`. `markAsEnum` won't be called more than once (thanks to the `isEnum()` check that was on line 397), but an earlier version of this code without that check didn't throw an exception, either, on our test cases. This version avoids re-computing the list of variable elements again when the class is already known to be an enum, which means the exception will not be thrown - but the stored enum values won't be modified, either.